### ISSUE TITLE
SSL_ERROR_WANT_WRITE error when using large certificate

### ISSUE THREAD

**Issue by jeffdunn** [grpc/grpc#17340]
### What version of gRPC and what language are you using?
1.14.1, C++
 
 
### What operating system (Linux, Windows, …) and version?
CentOS 7
 

### What runtime / compiler are you using (e.g. python version or version of gcc)
gcc 5.5.0
 
### What did you do?
I created a sample client/server which uses secure channels. I attempted to use existing certificates from our production stack, but found that depending on the size of the certificate, the server may fail with a protocol error.

This can be reproduced with a simple server and a generated certificate with many Subject Alternative Name entries.

**Repro steps:**
- Generate large certificate using [certstrap](https://github.com/square/certstrap)
    - Create [generate_domains.py](https://github.com/grpc/grpc/files/2630392/generate_domains.py.txt)
    - `./bin/certstrap init --common-name GrpcTestingCA`
    - `./bin/certstrap request-cert --domain full_hostname_goes_here.com,(python ./generate_domains.py 128) --uri (python ./generate_domains.py 58) --common-name GrpcTesting`
    - `./bin/certstrap sign GrpcTesting --CA GrpcTestingCA`
    - See [certs.tar.gz](https://github.com/grpc/grpc/files/2630516/certs.tar.gz) for sample certs from above steps 
- Run the following:
    - [GrpcClientMain.cpp](https://github.com/grpc/grpc/files/2630426/GrpcClientMain.cpp.txt)
    - [GrpcServerMain.cpp](https://github.com/grpc/grpc/files/2630431/GrpcServerMain.cpp.txt)
    - [HelloGrpcService.proto](https://github.com/grpc/grpc/files/2630434/HelloGrpcService.proto.txt)

### What did you expect to see?
A successful connection established with a request/response.
 
### What did you see instead?

When `GRPC_TRACE=all` is set:
```
GrpcServer grpc/protobuf running on port: 20133
E1129 10:43:02.077461452 2634068 lockfree_event.cc:94]       LockfreeEvent::NotifyOn: 0x610000004eb0 curr=(nil) closure=0x612000002de8
E1129 10:43:05.549992245 2634071 lockfree_event.cc:209]      LockfreeEvent::SetReady: 0x610000004eb0 curr=0x612000002de8
E1129 10:43:05.550497959 2634071 lockfree_event.cc:94]       LockfreeEvent::NotifyOn: 0x6100000080b0 curr=(nil) closure=0x616000120220
E1129 10:43:05.550521129 2634071 lockfree_event.cc:94]       LockfreeEvent::NotifyOn: 0x610000004eb0 curr=(nil) closure=0x612000002de8
E1129 10:43:05.550534468 2634071 lockfree_event.cc:209]      LockfreeEvent::SetReady: 0x6100000080b0 curr=0x616000120220
E1129 10:43:05.550540443 2634071 lockfree_event.cc:209]      LockfreeEvent::SetReady: 0x6100000080b8 curr=(nil)
E1129 10:43:05.553165626 2634071 ssl_transport_security.cc:1227] Handshake failed with fatal error SSL_ERROR_WANT_WRITE: error:00000000:lib(0):func(0):reason(0).
E1129 10:43:05.553240500 2634071 lockfree_event.cc:160]      LockfreeEvent::SetShutdown: 0x6100000080b0 curr=(nil) err={"created":"@1543516985.553180767","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
E1129 10:43:05.553283085 2634071 lockfree_event.cc:160]      LockfreeEvent::SetShutdown: 0x6100000080b8 curr=0x2 err={"created":"@1543516985.553180767","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
E1129 10:43:05.553293363 2634071 lockfree_event.cc:160]      LockfreeEvent::SetShutdown: 0x6100000080c0 curr=(nil) err={"created":"@1543516985.553180767","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
```

A snippet from when `GRPC_TRACE=all GRPC_VERBOSITY=debug` are both set:
```
I1129 10:42:35.777989899 2632278 ssl_transport_security.cc:211]      HANDSHAKE START -   before/accept initialization  - UNKWN
I1129 10:42:35.778053821 2632278 ssl_transport_security.cc:211]                 LOOP -   before/accept initialization  - UNKWN
I1129 10:42:35.778139255 2632278 ssl_transport_security.cc:211]                 LOOP -      SSLv3 read client hello A  - 3RCH_A
I1129 10:42:35.778152824 2632278 ssl_transport_security.cc:211]                 LOOP -     SSLv3 write server hello A  - 3WSH_A
I1129 10:42:35.778913595 2632278 ssl_transport_security.cc:211]                 LOOP -      SSLv3 write certificate A  - 3WSC_A
I1129 10:42:35.780631849 2632278 ssl_transport_security.cc:211]                 LOOP -     SSLv3 write key exchange A  - 3WSKEA
I1129 10:42:35.780666584 2632278 ssl_transport_security.cc:211]                 LOOP - SSLv3 write certificate reques  - 3WCR_A
E1129 10:42:35.780687823 2632278 ssl_transport_security.cc:1227] Handshake failed with fatal error SSL_ERROR_WANT_WRITE: error:00000000:lib(0):func(0):reason(0).
D1129 10:42:35.780728615 2632278 security_handshaker.cc:129] Security handshake failed: {"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
I1129 10:42:35.780754880 2632278 ev_posix.cc:222]            (fd-trace) fd_shutdown(8)
E1129 10:42:35.780782297 2632278 lockfree_event.cc:160]      LockfreeEvent::SetShutdown: 0x6100000080b0 curr=(nil) err={"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
E1129 10:42:35.780822583 2632278 lockfree_event.cc:160]      LockfreeEvent::SetShutdown: 0x6100000080b8 curr=0x2 err={"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
E1129 10:42:35.780833252 2632278 lockfree_event.cc:160]      LockfreeEvent::SetShutdown: 0x6100000080c0 curr=(nil) err={"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
I1129 10:42:35.780855777 2632278 handshaker.cc:212]          handshake_manager 0x61300000ffc0: error={"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"} shutdown=0 index=1, args={endpoint=(
nil), args=(nil) {size=0: (null)}, read_buffer=(nil) (length=0), exit_early=0}
I1129 10:42:35.780868352 2632278 handshaker.cc:245]          handshake_manager 0x61300000ffc0: handshaking complete -- scheduling on_handshake_done with error={"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"T
SI_PROTOCOL_FAILURE"}
I1129 10:42:35.780876264 2632278 timer_generic.cc:467]       TIMER 0x613000010040: CANCEL pending=true
D1129 10:42:35.780883863 2632278 chttp2_server.cc:113]       Handshaking failed: {"created":"@1543516955.780696841","description":"Handshake failed","file":"grpc/1.14.1/src/grpc-1.14.1/src/core/lib/security/transport/security_handshaker.cc","file_line":248,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
```

**Comment by ZhenLian** [comment]
We recently saw a similar issue internally, so reopen this.

**Comment by ZhenLian** [comment]
CC @yihuazhang @markdroth

**Comment by markdroth** [comment]
Was this fixed by #29176?

**Comment by matthewstevenson88** [comment]
This has been fixed through a combination of #29176 and #33638.
